crates/shared/src/price_estimation/trade_finder.rs

```diff
@@ -119,126 +148,211 @@ impl Inner {
     }
 }
 
+fn encode_interactions(interactions: &[Interaction]) -> Vec<EncodedInteraction> {
+    interactions.iter().map(|i| i.encode()).collect()
+}
+
+fn encode_settlement(query: &QuoteQuery, trade: &Trade, native_token: H160) -> EncodedSettlement {
+    let mut trade_interactions = encode_interactions(&trade.interactions);
+    if query.buy_token == BUY_ETH_ADDRESS {
+        // Because the `driver` manages `WETH` unwraps under the hood the `TradeFinder`
+        // does not have to emit unwraps to pay out `ETH` in a trade.
+        // However, for the simulation to be successful this has to happen so we do it
+        // ourselves here.
+        let buy_amount = match query.kind {
+            OrderKind::Sell => trade.out_amount,
+            OrderKind::Buy => query.in_amount,
+        };
+        let weth = dummy_contract!(WETH9, native_token);
+        let calldata = weth.methods().withdraw(buy_amount).tx.data.unwrap().0;
+        trade_interactions.push((native_token, 0.into(), Bytes(calldata)));
+        tracing::trace!("adding unwrap interaction for paying out ETH");
+    }
+
+    let tokens = vec![query.sell_token, query.buy_token];
+    let clearing_prices = match query.kind {
+        OrderKind::Sell => vec![trade.out_amount, query.in_amount],
+        OrderKind::Buy => vec![query.in_amount, trade.out_amount],
+    };
+
+    // Configure the most disadvantageous trade possible. Should the trader not
+    // receive the amount promised by the [`Trade`] the simulation will still work
+    // and we can compute the actual [`Trade::out_amount`] afterwards.
+    let (sell_amount, buy_amount) = match query.kind {
+        OrderKind::Sell => (query.in_amount, 0.into()),
+        OrderKind::Buy => (U256::MAX, query.in_amount),
```

Comment:
This will potentially overflows in the contract. I suggest `1 << 160` or something similar.